substrate/client/executor/polkavm/src/lib.rs

```diff
@@ -127,22 +101,31 @@ impl<'r, 'a> FunctionContext for Context<'r, 'a> {
 		dest: &mut [u8],
 	) -> sp_wasm_interface::Result<()> {
 		self.0
-			.read_memory_into_slice(u32::from(address), dest)
+			.instance
+			.read_memory_into(u32::from(address), dest)
 			.map_err(|error| error.to_string())
 			.map(|_| ())
 	}
 
 	fn write_memory(&mut self, address: Pointer<u8>, data: &[u8]) -> sp_wasm_interface::Result<()> {
-		self.0.write_memory(u32::from(address), data).map_err(|error| error.to_string())
+		self.0
+			.instance
+			.write_memory(u32::from(address), data)
+			.map_err(|error| error.to_string())
 	}
 
 	fn allocate_memory(&mut self, size: WordSize) -> sp_wasm_interface::Result<Pointer<u8>> {
-		let pointer = self.0.sbrk(0).expect("fetching the current heap pointer never fails");
-
-		// TODO: This will leak guest memory; find a better solution.
-		self.0.sbrk(size).ok_or_else(|| String::from("allocation failed"))?;
+		// Fetching the current heap pointer is always safe to unwrap.
+		let start = self.0.instance.sbrk(0).unwrap().unwrap();
+		let end = match self.0.instance.sbrk(size) {
+			Ok(Some(end)) => end,
+			Ok(None) => Err(String::from("#OOM (Out Of Memory)"))?,
+			Err(_) => Err(String::from("#GPE (General Purpose Error)"))?,
+		};
 
-		Ok(Pointer::new(pointer))
+		// Do a sanity check.
+		assert!(end - start == size);
+		Ok(Pointer::new(start))
```

Comment:
Restore the `expect` as it was, restore the comment about leaking memory, properly propagate the error it returns (`Err(error) => Err(format!("allocation failed: {error}"))`

Reply:
Yeah this was fixed based on compilation error so not sure if I can return it but I can check.

Reply:
https://github.com/paritytech/polkadot-sdk/pull/6533/commits/f173aae5fedbd5c68af96cb980728662b9d6c623

Reply:
https://github.com/paritytech/polkadot-sdk/pull/6533/commits/2590aab278f2acbc9ee3db9466ae071e13272937